gc.c

```diff
@@ -1517,42 +1494,55 @@ check_rvalue_consistency_force(const VALUE obj, int terminate)
             err++;
         }
         else if (!is_pointer_to_heap(objspace, (void *)obj)) {
-#ifndef USE_THIRD_PARTY_HEAP
-            /* check if it is in tomb_pages */
-            struct heap_page *page = NULL;
-            for (int i = 0; i < SIZE_POOL_COUNT; i++) {
-                rb_size_pool_t *size_pool = &size_pools[i];
-                ccan_list_for_each(&size_pool->tomb_heap.pages, page, page_node) {
-                    if (page->start <= (uintptr_t)obj &&
-                            (uintptr_t)obj < (page->start + (page->total_slots * size_pool->slot_size))) {
-                        fprintf(stderr, "check_rvalue_consistency: %p is in a tomb_heap (%p).\n",
-                                (void *)obj, (void *)page);
-                        err++;
-                        goto skip;
+#if USE_MMTK
+            if (!mmtk_enable) {
+#endif
+                /* check if it is in tomb_pages */
+                struct heap_page *page = NULL;
+                for (int i = 0; i < SIZE_POOL_COUNT; i++) {
+                    rb_size_pool_t *size_pool = &size_pools[i];
+                    ccan_list_for_each(&size_pool->tomb_heap.pages, page, page_node) {
+                        if (page->start <= (uintptr_t)obj &&
+                                (uintptr_t)obj < (page->start + (page->total_slots * size_pool->slot_size))) {
+                            fprintf(stderr, "check_rvalue_consistency: %p is in a tomb_heap (%p).\n",
+                                    (void *)obj, (void *)page);
+                            err++;
+                            goto skip;
+                        }
                     }
                 }
-            }
+#if USE_MMTK
+    }
 #endif
             bp();
             fprintf(stderr, "check_rvalue_consistency: %p is not a Ruby object.\n", (void *)obj);
             err++;
-#ifndef USE_THIRD_PARTY_HEAP
           skip:
-#endif
             ;
         }
         else {
-#ifndef USE_THIRD_PARTY_HEAP
-            // TODO remove these
-            const int wb_unprotected_bit = RVALUE_WB_UNPROTECTED_BITMAP(obj) != 0;
-            const int uncollectible_bit = RVALUE_UNCOLLECTIBLE_BITMAP(obj) != 0;
-            const int mark_bit = RVALUE_MARK_BITMAP(obj) != 0;
-            const int marking_bit = RVALUE_MARKING_BITMAP(obj) != 0, remembered_bit = marking_bit;
-            const int age = RVALUE_FLAGS_AGE(RBASIC(obj)->flags);
-
-            if (GET_HEAP_PAGE(obj)->flags.in_tomb) {
-                fprintf(stderr, "check_rvalue_consistency: %s is in tomb page.\n", obj_info(obj));
-                err++;
+            int wb_unprotected_bit;
+            int uncollectible_bit;
+            int mark_bit;
+            int marking_bit;
+            int remembered_bit;
+            int age;
+
+#if USE_MMTK
+            if (!mmtk_enable) {
```

Comment:
I managed to workaround this by assigning the result to a local variable.  The compiler seems smart enough to analyse the control flow w.r.t. local variables.

```suggestion
            const bool mmtk_enabled_local = rb_mmtk_enabled_p(); 
            if (!mmtk_enabled_local) {
```